### PR TITLE
add ActiveInterfaceOnly option to interface plugin

### PR DESCRIPTION
**Pull request by rinigus**
This PR adds ActiveInterfaceOnly to interface plugin. 

Rationale: In Sailfish devices there are very many network interfaces (28 in Nexus 4) out of which only 3 are actually used: one for cellular data, one for WIFI, and lo. Unfortunately, active interfaces on different devices could be different leading to inability to enable only the active ones using Interface option for all devices in default configuration. So, users would have to find out which interface is active for them and enable these interfaces manually.

By adding ActiveInterfaceOnly option, only interfaces with packet count larger than 0 would be reported. The implementation has been tested on Linux (Sailfish), but I hope that it works on other platforms as well.

### CODE REVIEW

**Comment by octo** [comment]
Hi @rinigus, thank you very much for this patch. I really like the functionality, in fact I've written [a similar plugin](https://collectd.org/wiki/index.php/Match:Empty_Counter) a while ago. I'd consider that plugin currently broken, though, since it *only* works with `COUNTER` metrics and those are basically unused with the default plugins.

I'll leave some review comments inline, but from a quick glance these will likely mostly be about coding style, i.e. easily fixable.

Best regards,
—octo

File: src/collectd.conf.in [hunk 1]
```diff
@@ -582,6 +582,7 @@
 #<Plugin interface>
 #	Interface "eth0"
 #	IgnoreSelected false
+#	ActiveInterfaceOnly false
```
Comment (octo) [inline src/collectd.conf.in]: I'm a bit unhappy about the option name. How do you feel about *ReportInactive*?

File: src/interface.c [hunk 2]
```diff
@@ -84,12 +84,15 @@ static const char *config_keys[] =
 {
 	"Interface",
 	"IgnoreSelected",
+	"ActiveInterfaceOnly",
 	NULL
 };
-static int config_keys_num = 2;
+static int config_keys_num = 3;
```
Comment (octo) [inline src/interface.c]: The *interface plugin* is very old and this line is one of the lines where it shows. Please remove the `NULL` from the `config_keys` array and use:
```c
static int config_keys_num = STATIC_ARRAY_SIZE (config_keys);
```

File: src/interface.c [hunk 3]
```diff
@@ -114,6 +117,13 @@ static int interface_config (const char *key, const char *value)
 			invert = 0;
 		ignorelist_set_invert (ignorelist, invert);
 	}
+	else if (strcasecmp (key, "ActiveInterfaceOnly") == 0)
+	{
+		if (IS_TRUE (value))
```
Comment (octo) [inline src/interface.c]:
```c
active_interface_only = IS_TRUE (value);
```

File: src/interface.c [hunk 4]
```diff
@@ -223,6 +233,9 @@ static int interface_read (void)
 		if (if_ptr->ifa_addr != NULL && if_ptr->ifa_addr->sa_family == AF_LINK) {
 			if_data = (struct IFA_DATA *) if_ptr->ifa_data;
 
+			if ( active_interface_only && if_data->IFA_RX_PACKT == 0 && if_data->IFA_TX_PACKT == 0 )
```
Comment (octo) [inline src/interface.c]: Please remove extra spaces within the parenthesis. Here and below.

File: src/collectd.conf.pod [hunk 5]
```diff
@@ -2735,6 +2735,18 @@ This will ignore the loopback interface, all interfaces with names starting
 with I<veth> and all interfaces with names starting with I<tun> followed by
 at least one digit.
 
+=item B<ActiveInterfaceOnly> I<true>|I<false>
```
Comment (octo) [inline src/collectd.conf.pod]: Please change to `I<false>|I<true>`, so that the default comes first.

**Comment by rinigus** [comment]
Excellent ! I'll insert the changes and update PR. For background, can't we change counter to derive? I think that makes more sense than counter. or do I miss something?

best wishes,

rinigus

**Comment by rinigus** [comment]
@octo: thank you for review and suggestions. I think I managed to put them all in.

best wishes,

rinigus

File: src/interface.c [hunk 6]
```diff
@@ -426,3 +426,13 @@ void module_register (void)
...
+
+/*
+ * Local variables:
+ *  c-file-style: "linux"
+ *  indent-tabs-mode: t
+ *  c-indent-level: 4
+ *  c-basic-offset: 4
+ *  tab-width: 4
+ * End:
+ */
```
Comment (rubenk) [inline src/interface.c]: I know there's prior art all over the place in the code base, and please know that this is nothing personal, but I think these mode lines are distracting at least and don't help at all. Most editors should be able to set per-project settings these days.

**Comment by rinigus** [comment]
Hi @rubenk 

I don't take it personal - no worries. If you don't want to see the formatting specs in each file then I think you should alter the https://collectd.org/wiki/index.php/Coding_style . As I read the Coding style document, each file has (potentially) its own style. So, there are no project specific formatting instructions. 

As far as I have seen the parts I looked into, most of the code is using tabs as tabs (not spaces) and Linux-style braces. Maybe this should be then required for new code. That would allow everyone select tab width as they like and get on with it without breaking much. 

Please make a decision regarding formatting and if you move towards unified format requirements.

@rubenk and @octo: I'll wait for your reply before removing/altering this PR.

**Comment by rubenk** [comment]
I'd thought the first 2 points in the Coding Style cover it al:
- If you're starting a new file, we recommend to indent with two spaces.
- If you're writing a patch, please use the indentation of the surrounding code.

These recommendations are not always followed though unfortunately, and the document itself is a bit hard to find.

I have no preference for tabs or spaces, but like to avoid a mix of them.
Have a look at https://github.com/collectd/collectd/pull/1822/commits/30bb3da506ce9ab8b03316cc71263239c1e2e40d and you'll see what I mean. The nesting became so hard to follow it cost me an hour to see the actual issue I was looking for in that file.

If we all agree, I can add instructions about the bracing style, and about not adding more mode lines and fold markers.

**Comment by rinigus** [comment]
I don't mind having more specific formatting instructions, like not adding mode lines and bracing style. As the code grows, it is probably good to have more restrictions on the style - this would allow everyone to spend less time on it. 

Use of two spaces for indentation is somewhat conflicting with the current code, as I saw it. To me it looks that tabs are used a lot and maybe its for the better - everyone can select how they see these tabs in the editor through editor settings.

I'll proceed with this PR by removing mode lines

**Comment by rinigus** [comment]
mode lines removed from the end of the plugin code

**Comment by rubenk** [comment]
I agree, we need more instructions and stricter policies as we grow.

As far as I can tell, all the original code written by @octo was indented by 2 spaces. Contributed files differ in style a lot. Unfortunately mass reformatting all of them creates a lot of churn, makes it hard to follow history and do backports.

As for which style is better, let's not have a discussion about tabs vs spaces on Sunday morning.
Or switch to Go, gofmt all the things and be done with it ;)

**Comment by rinigus** [comment]
:). OK, lets' forget about tabs/spaces and get back to the PR.

I don't know much about original version, but the file that I pulled from git and patched has been written as it was. If you see the diff, the changes are only on the lines which were changed, not reformatted. There are no changes with whitespace/tabs. 

I had to shift some code to implement the feature (like getting packets before octets), but to my knowledge, I tried to follow the same style as the code was before I pulled it.

**Comment by octo** [comment]
LGTM. Personally I don't mind these mode lines, especially since the plugins' coding style is so all-over-the-place.

Coding style wise, the most important is **don't surprise the reader**, which in this case translates to "blend in with your surroundings" – the *interface plugin* uses tabs, so use tabs for your additional code, too. Try to mimick spacing, parenthesis and bracing style, etc. pp.

Unfortunately, as a free project we have to strike a reasonable balance between maintaining coding style and scaring off contributors. If we get too nit-picky, we're just annoying contributors and the round trip times are high enough as it is. For example, you open a brace at the end of line 366 while the rest of the file puts opening braces on a line of their own. I *could* pick on this, but I don't feel it's worth the back and forth.

Originally *collectd* used 8-wide tabs, eventually we started pushing for 2 spaces instead. The *interface plugin* is old enough that it is using tabs still.

I agree with the *gofmt* approach, i.e. automatically reformat before each commit. I hear good things about [ClangFormat](http://clang.llvm.org/docs/ClangFormat.html) but the old and tested *indent* might also be an option. The annoying thing about reindenting (and reason why I haven't so already) is that it makes porting patches between branches incredibly hard, i.e. a user writes a patch, we have to back-port it to a version branch (e.g. *collectd-5.5*) and then we're guaranteed merge conflicts when merging *collectd-5.4* into *master*.

Alright, long rambling short: code looks fine, let's pull it and have a discussion about (automatic) code formatting on the mailing list.

**Comment by rinigus** [comment]
@octo and @rubenk : Thank you for explanations and time you took for writing them! And indeed, the formatting of the code could be discussed in the mailing list. Just want to thank you for your work and, hopefully, its possible to find the solution that makes your jobs easier.
